modules/mastodon/index.js

```diff
@@ -0,0 +1,288 @@
+          fn(new Error('record is not a Note or Person'));
+        } else {
+          fn(null, record);
+        }
+      } catch (e) {
+        fn(e);
+      }
+    }
+  });
+}
+
+function htmlToText(s) {
+  return htmlToTextMod.fromString(s, {
+    wordwrap: null,
+    ignoreHref: true,
+  });
+}
+
+function formatRecord(record, fn) {
+  if (record.type == 'Note') {
+    fetchResource(record.attributedTo, (error, userRecord) => {
+      var str = '';
+      if (error) {
+        str += '(error getting user): ';
+      } else {
+        str += sprintf('%s (@%s): ', userRecord.name, userRecord.preferredUsername);
+      }
+      if (record.sensitive) {
+        str += sprintf('[content warning] %s', record.summary);
+      } else {
+        str += htmlToText(record.content);
+      }
+      fn(str);
+    });
+  } else if (record.type == 'Person') {
+    fn(sprintf('%s (@%s): %s',
+      record.name,
+      record.preferredUsername,
+      htmlToText(record.summary),
+    ));
+  }
+}
+
+module.exports.commands = ['masto'];
+
+module.exports.init = function(b) {
+  bot = b;
+  log = bot.log;
+
+  bot.readDataFile('config.json', function(err, data) {
+    if (err) {
+      log.info('initializing mastodon module data');
+      moduleConfig = defaultConfig();
+    } else {
+      try {
+        moduleConfig = JSON.parse(data);
+      } catch(ex) {
+        moduleConfig = defaultConfig();
+      }
+    }
+    writeConfig();
+  });
+
+  var userAgent = moduleConfig && moduleConfig.userAgent || 'EuIrcBot';
+  try {
+    var ver = require('../../package.json').version;
+    userAgent += '/' + ver;
+  } catch (ex) {}
+
+  var userAgentUrl = (moduleConfig && moduleConfig.userAgentUrl
+    || 'https://github.com/euank/EuIrcBot/blob/{commitish}/modules/mastodon/info.md');
+  if (userAgentUrl.includes('{commitish}')) {
+    var commitish = 'master';
+    try {
+      commitish = execSync('git rev-parse HEAD', {'encoding': 'utf8'}).trim();
+    } catch (ex) {}
+    userAgentUrl = userAgentUrl.replace(/{commitish}/g, commitish);
+  }
+  userAgent += ' (+' + userAgentUrl + ')';
+
+  request = require('request').defaults({
+    headers: {'User-Agent': userAgent}
+  });
+};
+
+/*
+ * When a user calls "!masto [url]", both run and url get called. The first
+ * thing we do in each of these methods is check the domain against
+ * knownInstances; run only continues if the instance is unknown, and url only
+ * continues if the instance is known.
+ */
+
+function handleUrl(url, reply) {
+  fetchResource(url, (error, record) => {
+    if (error) {
+      error.mastodonUrl = url;
+      log.info(error);
+      reply('error: ' + error.message);
+    } else {
+      formatRecord(record, (s) => reply.custom({ replaceNewlines: true }, s));
+    }
+  });
+}
+
+module.exports.run = function(remainder, parts, reply, command, from, to, text, raw) {
+  url = urlMod.parse(remainder);
+  checkHostname(url.host, (ok) => { if (!ok) {
```

Comment:
As I read this, in the happy case of `checkHostname` `handleUrl` doesn't end up getting invoked; am I missing something?